Source/OxyPlot/Series/ContourSeries.cs

```diff
@@ -277,26 +285,98 @@ public override void Render(IRenderContext rc)
+                        quotient = contour.ContourLevel / factoredStep;
+                    }
+                }
+
+                // don't add label to contours, if ContourLevel is not close to n * LabelStep
+                if (transformedPoints.Length <= 10 || Math.Abs(quotient - 1) >= 0.1) continue;
+
+                if (!this.MultiLabel)
+                {
+                    this.AddContourLabels(contour, transformedPoints, clippingRect, contourLabels, (transformedPoints.Length - 1) * 0.5);
+                    continue;
+                }
+
+                //calculate how many labels fit per contour
+                int labelsCount = (int)(contourLength / this.LabelSpacing);
+                if (labelsCount == 0)
+                {
+                    this.AddContourLabels(contour, transformedPoints, clippingRect, contourLabels, (transformedPoints.Length - 1) * 0.5);
+                    continue;
+                }
+
+                double contourPartLength = 0;
+                double contourPartLengthOld = 0;
+                int intervalIndex = 1;
+                double contourPartLengthTarget = 0;
+                double contourFirstPartLengthTarget = (contourLength - (labelsCount - 1) * this.LabelSpacing) / 2;
+                for (int j = 0; j < labelsCount; j++)
                 {
-                    var transformedPoints = contour.Points.Select(this.Transform).ToArray();
-                    var strokeColor = contour.Color.GetActualColor(this.ActualColor);
-
-                    rc.DrawClippedLine(
-                        clippingRect,
-                        transformedPoints,
-                        4,
-                        this.GetSelectableColor(strokeColor),
-                        this.StrokeThickness,
-                        this.EdgeRenderingMode,
-                        dashArray,
-                        LineJoin.Miter);
-
-                    // rc.DrawClippedPolygon(transformedPoints, clippingRect, 4, model.GetDefaultColor(), OxyColors.Black);
-                    if (transformedPoints.Length > 10)
+                    double i = 0;
+
+                    if (intervalIndex == 1)
+                    {
+                        contourPartLengthTarget = contourFirstPartLengthTarget;
+                    }
+                    else
                     {
-                        this.AddContourLabels(contour, transformedPoints, clippingRect, contourLabels);
+                        contourPartLengthTarget = contourFirstPartLengthTarget + j * this.LabelSpacing;
                     }
+
+                    //find index of contour points where next label should be positioned
+                    for (int k = intervalIndex; k < transformedPoints.Length; k++)
+                    {
+                        contourPartLength += Math.Sqrt(Math.Pow(transformedPoints[k].x - transformedPoints[k - 1].x, 2) + Math.Pow(transformedPoints[k].y - transformedPoints[k - 1].y, 2));
```

Comment:
Could be 

                        contourPartLength += (transformedPoints[k] - transformedPoints[k - 1]).Length;

For clarity